packages/beacon-node/src/chain/chain.ts

```diff
@@ -935,6 +939,28 @@ export class BeaconChain implements IBeaconChain {
     if (headState) {
       this.opPool.pruneAll(headState, finalizedState);
     }
+
+    // TODO: finalizedState may not be available after https://github.com/ChainSafe/lodestar/issues/5968.
+    // In that case implement getCheckpointStateValidatorCount() in regen and use headState instead
+    if (finalizedState === null) {
```

Comment:
Can you research ways to drop this dependency on the finalized state to do the key pruning? What happens if the pruning is never done because the finalized state is not available?

Reply:
I was looking into `activation_eligibility_epoch` to decide the cutoff but it's not reliable since a long chunk of partial deposits would prevent keys from moving into the finalized section.

Instead you can add a new cache into the state tracking historical validator length counts. On every epoch transition push the current count of validators. Once a state finalizes, retrieve count at index `state.epoch - finalized_checkpoint.epoch`. Then you don't need to retrieve the finalized state for this purpose.

Reply:
> I was looking into activation_eligibility_epoch to decide the cutoff but it's not reliable since a long chunk of partial deposits would prevent keys from moving into the finalized section.

Yea, a finalized validator doesn’t necessarily mean it is an active validator.